### PR TITLE
[FLINK-7368][metrics] Make MetricStore ThreadSafe class

### PR DESCRIPTION
**Pull request by pnowojski**
Remove external synchronisation on MetricStore

## What is the purpose of the change

This is a refactor that makes `MetricStore` thread safe. It achieves it by pulling in all modifying methods and by returning immutable copies in getters.

## Verifying this change

This change is already covered by existing tests, such as `MetricStoreTest` or `MetricFetcherTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java [hunk 1]
```diff
@@ -38,29 +43,136 @@
...
+	private final ComponentMetricStore jobManager = new ComponentMetricStore();
+	private final Map<String, TaskManagerMetricStore> taskManagers = new ConcurrentHashMap<>();
+	private final Map<String, JobMetricStore> jobs = new ConcurrentHashMap<>();
+
+	/**
+	 * Remove not active task managers.
+	 *
+	 * @param activeTaskManagers to retain.
+	 */
+	public synchronized void retainTaskManagers(List<String> activeTaskManagers) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: Isn't the method synchronization unnecessary now?

Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: Strictly speaking right now it isn't. However without synchronising those methods, implementing them is much harder. For example I made one mistake where I implemented:
```
if (!jobs.containKey(jobID)) {
  return null;
}
return jobs.get(jobID).getTaskMetricStore(taskID)
```
Also without synchronising it's harder to understand correctness of code, that's handling three separate fields (whether this is a correct access order; whether the state of variable is/should be coherent, like if existence of key in one map, should imply existence a key on another, etc). 

Long story short, synchronising is not an issue here, but helps with reasoning about this code.

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: but now we're paying the synchronization costs twice, once for the synchronized keyword and once again for every access to the map. If every method is synchronized the maps don't have to be concurrent hash maps.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java [hunk 2]
```diff
@@ -260,50 +293,66 @@ public String getMetric(String name, String defaultValue) {
...
+		public TaskManagerMetricStore(Map<String, String> metrics, Set<String> garbageCollectorNames) {
+			super(metrics);
+			this.garbageCollectorNames = checkNotNull(garbageCollectorNames);
+		}
 
 		public void addGarbageCollectorName(String name) {
 			garbageCollectorNames.add(name);
 		}
+
+		public static TaskManagerMetricStore unmodifiable(TaskManagerMetricStore source) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: I'm wondering whether we really need this unmodifiable business. Yes, it's technically a good idea, but the access to the MetricStore is limited and fully under our control; so we _know_ that we never try to modify the map.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java (hunk 1 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: 1. performance here is not that big a deal, but code correctness is. With `synchronized` it is just easier to implement any changes here in a thread safe manner. Without it, any new developer coming here will have to understand much more assumptions about this code, like whether consistency matters here or nor? whether order of the operations/access to the fields is important or nor? etc...

2. even with `synchronized` we still need either concurrent hash maps, because we return and make them visible to the outside world by getters. So we either need concurrent hash maps or return copies of them.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java (hunk 2 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: It's a matter of clean design and maintaining it in the future. You only *know* that we do not modify this map **at the moment**. Are you sure that every committer reviewing changes in this parts of the code will always spot any such modifications without `unmodifiable` assertion?  On the other hand removing `unmodifiable` in a PR is big yellow warning light that's something is going on.

Following this logic, why don't we make all methods/fields `public`, since we know that we will never try to access supposed to be `private` fields?

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: In which case we should rather modify the store to not allow writes in the first place, instead of opting for unmodifiable collections that are pretty much a hack. "here, have an object that fails for 50% of the defined methods"; that's hardly good design is it.

Till suggested dedicated read methods that return metrics, something like `List<String> getMetrics(List<String> metricNames)` instead of exposing the metric maps.

This would make the interface cleaner and would allow us to simplify the synchronization.

Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: Would you rather have a runtime error about illegal/unintended write access or none at all? Java sucks in this regard (missing const-correctness) :(

Exposing `List<...>` have the same issue.

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: If the returned list is a fresh copy independent of the map it can be modifiable; the user can do whatever he wants with it.

Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: Then you pay costs of copying all of the objects, which is much higher compared to the synchronisation. But yes, you can do that. In that case you don't need concurrent hash maps. 

If you intend to go in other direction with this and leave the current code in the master as it is, please log the work in Jira for this after the release (I will close release blocker issue for which I started this PR). However I have an impression that any follow up work would be easier on top of this change and in the mean time this is still way better then current code on master branch with this external synchronisation.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/util/MutableIOMetrics.java [hunk 3]
```diff
@@ -79,7 +79,7 @@ public void addIOMetrics(AccessExecution attempt, @Nullable MetricFetcher fetche
 				fetcher.update();
 				MetricStore metricStore = fetcher.getMetricStore();
 				synchronized (metricStore) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/util/MutableIOMetrics.java]: leftover synchronization

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java [hunk 4]
```diff
@@ -38,29 +43,136 @@
...
+	public synchronized void retainTaskManagers(List<String> activeTaskManagers) {
+		taskManagers.keySet().retainAll(activeTaskManagers);
+	}
+
+	/**
+	 * Remove not active task managers.
+	 *
+	 * @param activeJobs to retain.
+	 */
+	public synchronized void retainJobs(List<String> activeJobs) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: can be package private

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java [hunk 5]
```diff
@@ -38,29 +43,136 @@
...
+		jobs.keySet().retainAll(activeJobs);
+	}
+
+	/**
+	 * Add metric dumps to the store.
+	 *
+	 * @param metricDumps to add.
+	 */
+	public synchronized void addAll(List<MetricDump> metricDumps) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: package private

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java [hunk 6]
```diff
@@ -38,29 +43,136 @@
...
+	}
 
 	// -----------------------------------------------------------------------------------------------------------------
-	// Adding metrics
+	// Accessors for sub MetricStores
 	// -----------------------------------------------------------------------------------------------------------------
-	public void add(MetricDump metric) {
+
+	/**
+	 * Returns the {@link ComponentMetricStore}.
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: javadoc is missing any mention of jobmanager

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java [hunk 7]
```diff
@@ -38,29 +43,136 @@
...
+	 *
+	 * @return JobManagerMetricStore
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: outdated class reference

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java [hunk 8]
```diff
@@ -181,74 +265,23 @@ private void addMetric(Map<String, String> target, String name, MetricDump metri
...
-	 *
-	 * @param jobID  job ID
-	 * @param taskID task ID
-	 * @return TaskMetricStore for given IDs, or null if no store for the given arguments exists
-	 */
-	public TaskMetricStore getTaskMetricStore(String jobID, String taskID) {
-		JobMetricStore job = getJobMetricStore(jobID);
-		if (job == null) {
-			return null;
+		public ComponentMetricStore() {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: all constructors can be private

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java [hunk 9]
```diff
@@ -260,50 +293,66 @@ public String getMetric(String name, String defaultValue) {
 				? value
 				: defaultValue;
 		}
-	}
 
-	/**
-	 * Sub-structure containing metrics of the JobManager.
-	 */
-	public static class JobManagerMetricStore extends ComponentMetricStore {
+		public static ComponentMetricStore unmodifiable(ComponentMetricStore source) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java]: can be private

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStoreTest.java [hunk 10]
```diff
@@ -58,9 +59,9 @@ public void testMalformedNameHandling() {
 		store.add(cd);
 
 		//-----verify that no side effects occur
-		assertEquals(0, store.jobManager.metrics.size());
-		assertEquals(0, store.taskManagers.size());
-		assertEquals(0, store.jobs.size());
+		assertTrue(store.getTaskManagers().isEmpty());
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStoreTest.java]: i actually prefer the previous version as it prints the actual value.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/util/MutableIOMetrics.java (hunk 3 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/util/MutableIOMetrics.java]: Don't know why I have missed it :(

**Comment by pnowojski** [comment]
Thanks!
